### PR TITLE
[[ Bug 15565 ]] Add missing create syntax forms

### PR DESCRIPTION
**Pull request by abrisan**
I have added the following (i.e. these are now accepted statements)

create card .. in stack ...
create control ... in card ... 

where the stack and card are not the default stack/card

### CODE REVIEW

File: engine/src/card.h [hunk 1]
```diff
@@ -136,7 +136,7 @@ class MCCard : public MCObject, public MCMixinObjectHandle<MCCard>
 	IO_stat loadobjects(IO_handle stream, uint32_t version);
 
 	void kfocusset(MCControl *target);
-	MCCard *clone(Boolean attach, Boolean controls);
+	MCCard *clone(Boolean attach, Boolean controls, MCStack *p_parent=nullptr);
```
Comment (runrevmark) [inline engine/src/card.h]: Missing spaces around '='

Comment (abrisan) [reply engine/src/card.h]: Fixed

File: engine/src/exec-interface.cpp [hunk 2]
```diff
@@ -3299,7 +3299,7 @@ void MCInterfaceExecCreateCard(MCExecContext& ctxt, MCStringRef p_new_name, MCSt
     ctxt . SetItToValue(*t_id);
 }
 
-MCControl* MCInterfaceExecCreateControlGetObject(MCExecContext& ctxt, int p_type, MCGroup *&r_parent)
+MCControl* MCInterfaceExecCreateControlGetObject(MCExecContext& ctxt, int p_type, MCObject *&r_parent)
```
Comment (runrevmark) [inline engine/src/exec-interface.cpp]: I don't think the r_parent var is used - if not, we can remove it.

File: docs/dictionary/command/create-card.lcdoc [hunk 3]
```diff
@@ -24,10 +24,16 @@ create card "Main"
...
+Example:
+create card "Main" in stack "Stack"
+
 Parameters:
 cardName:
 The name property of the new card. If you don't specify a cardName, the
 new card's name is empty.
+stackName:
+The stack in which the new card is placed. If you don't specify
+a stackName, it is placed in defaultStack.
```
Comment (livecodeali) [inline docs/dictionary/command/create-card.lcdoc]: This is probably better as `in the defaultStack`. Also, if you put `<defaultStack>` that creates a link to the defaultStack entry in the docs.

File: docs/dictionary/command/create-card.lcdoc [hunk 4]
```diff
@@ -37,16 +43,13 @@ Description:
 Use the <create card> <command> to add a new <card> to the <current
 stack>. 
 
-The new card is placed after the current card in the defaultStack. When
-the <card> is created, you go to the new <card>.
+The new card is placed after the current card. If <stackName>
+is specified, it will be placed in <stackName>, otherwise 
+in defualtStack When the <card> is created, you go to the new <card>.
```
Comment (livecodeali) [inline docs/dictionary/command/create-card.lcdoc]: `... in the <defaultStack>. When..`

File: engine/src/exec-interface.cpp [hunk 5]
```diff
@@ -3330,7 +3330,19 @@ MCControl* MCInterfaceExecCreateControlGetObject(MCExecContext& ctxt, int p_type
 
 void MCInterfaceExecCreateControl(MCExecContext& ctxt, MCStringRef p_new_name, int p_type, MCObject *p_container, bool p_force_invisible)
 {
-    if (MCdefaultstackptr->islocked())
+    
+    MCStack *t_current_stack;
```
Comment (runrevmark) [inline engine/src/exec-interface.cpp]: We try not to leave POD-typed vars (those which do not have a constructor) uninit'd on declaration. I'd suggest:
```
MCStack *t_current_stack = MCdefaultstackptr;
if (p_container != nullptr)
{
  t_current_stack = p_container->getstack();
}
```

File: tests/lcs/core/interface/stack.livecodescript [hunk 6]
```diff
@@ -47,4 +47,22 @@ on TestCreateControlInStack
 
 
 	TestAssert "test if the card was created in the correct stack" , there is a card "Other" of stack "Temp"
-end TestCreateControlInStack
\ No newline at end of file
+end TestCreateControlInStack
+
+on TestStackLockedness
```
Comment (runrevmark) [inline tests/lcs/core/interface/stack.livecodescript]: I suggest TestCreateControlInContainerWithDefaultStackLocked

File: docs/dictionary/command/create.lcdoc [hunk 7]
```diff
@@ -4,7 +4,7 @@ Synonyms: new
 
 Type: command
 
-Syntax: create [invisible] <objectType> [<objectName>] [in <group>]
+Syntax: create [invisible] <objectType> [<objectName>] [in <group>] [in <card>]
```
Comment (livecodeali) [inline docs/dictionary/command/create.lcdoc]: This should be either/or, so 
`[in {group | card}]`

**Comment by runrevmark** [comment]
@alexbrisan The same bug is present in CreateWidget too - they are the same bug, so fix them as part of 19703.

File: docs/dictionary/command/create.lcdoc [hunk 8]
```diff
@@ -40,8 +43,13 @@ group:
 A reference or and expression that evaluates to a reference to a
 group on the current card. If you specify a group, the new object is a 
 member of the group, and exists on each card that has the group. If you 
-don't specify a group, the object is created on the current card and 
-appears only on that card.
+don't specify either a group nor a card, the object is created on the
```
Comment (livecodeali) [inline docs/dictionary/command/create.lcdoc]: maybe `specify neither a group nor a card`, rather than `don't specify either...`

**Review by livecodeali — CHANGES_REQUESTED** [review]

File: docs/dictionary/command/create-card.lcdoc [hunk 9]
```diff
@@ -4,7 +4,7 @@ Synonyms: new card
 
 Type: command
 
-Syntax: create card [<cardName>]
+Syntax: create card [<cardName>] [in stack <stackName>]
```
Comment (runrevmark) [inline docs/dictionary/command/create-card.lcdoc]: The additional syntax should be `[in <stack>]`

File: docs/dictionary/command/create-card.lcdoc [hunk 10]
```diff
@@ -24,10 +24,16 @@ create card "Main"
...
 create card (field "ouput")
 
+Example:
+create card "Main" in stack "Stack"
+
 Parameters:
 cardName:
 The name property of the new card. If you don't specify a cardName, the
 new card's name is empty.
+stackName:
```
Comment (runrevmark) [inline docs/dictionary/command/create-card.lcdoc]:
```
stack:
...
```
Use a similar description as for <group> in the other doc.

**Comment by montegoulding** [comment]
@alexbrisan This is going to be _really_ helpful! 💯 

Bikeshed: I suspect I would more naturally write `on card | stack` instead of `in` as with groups. Of course it's probably better to stick to one form and we already use `in`.

File: docs/dictionary/command/create.lcdoc (hunk 7 above)
Comment (livecodeali) [reply docs/dictionary/command/create.lcdoc]: Sorry, I gave you bad advice here - it needs to be 
`[in {<group> | <card>}]` as things that are parameters to the syntax have to be enclosed in `<>`

**Comment by livecodeali** [comment]
@alexbrisan Two additional things I've just noticed! (sorry)
- The `MCGroup *` cast should be removed from `MCInterfaceExecCreateControl(ctxt, *t_new_name, otype, (MCGroup *)parent, visible == False);`
- `MCInterfaceExecCreateWidget(ctxt, *t_new_name, otype, (MCGroup *)parent, visible == False);` should also be able to take a card _or_ group reference!

**Review by livecodeali — APPROVED** [review]

**Comment by livecodeali** [comment]
@livecode-vulcan review ok 48f8ae0

**Comment by livecode-vulcan** [comment]
💙 review by @livecodeali ok 48f8ae021c812d94b543c1d85ed28570d3b1f548

**Comment by livecode-vulcan** [comment]
😞 test failure 48f8ae021c812d94b543c1d85ed28570d3b1f548

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/254)
* try-community-armv6-android-api9: [failure](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/263)
* try-community-js-emscripten-sdk1.35: [failure](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/243)
* try-community-universal-ios-iphoneos10.3: [failure](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/186)
* try-community-universal-ios-iphonesimulator10.3: [failure](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/186)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/252)
* try-community-universal-mac-macosx10.9: [failure](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/257)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/256)
* try-community-x86-linux-debian8: [failure](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/261)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/256)
* try-community-x86_64-linux-debian8: [failure](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/261)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/280)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/251)

**Comment by runrevmark** [comment]
@alexbrisan : Could you add a test for 'create widget' too - as the codepaths are different in the engine... Hopefully @livecodeali can advise how to make sure the test environment has a widget to try and create :)

File: tests/lcs/core/interface/stack.livecodescript [hunk 11]
```diff
@@ -39,12 +39,41 @@ on TestCreateControlInCard
...
+
+on TestCreateWidgetInCard
+	create stack "Temp"
+	
+	create card "Other"
+	create card "Other 2"
+
+	create widget "My Navbar" as "com.livecode.widget.navbar" in card "Other"
+
+	TestAssert "test if the widget was created in the correct card", there is a widget "My Navbar" of card "Other"
```
Comment (runrevmark) [inline tests/lcs/core/interface/stack.livecodescript]: use 'of stack "Temp"' here to make sure it is entirely unambiguous

**Comment by runrevmark** [comment]
@livecode-vulcan review ok 1af86eb

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok 1af86ebfa128ce62722c1385bcf91729349b907c

**Comment by livecode-vulcan** [comment]
😞 test failure 1af86ebfa128ce62722c1385bcf91729349b907c

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/269)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/278)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/258)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/201)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/201)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/267)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/272)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/271)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/276)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/271)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/276)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/295)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/266)

**Comment by abrisan** [comment]
@livecode-vulcan retry 1af86eb

**Comment by abrisan** [comment]
@livecode-vulcan retry 1af86ebfa128ce62722c1385bcf91729349b907c

**Comment by livecodepanos** [comment]
@livecode-vulcan retry 1af86eb

**Comment by livecode-vulcan** [comment]
😞 test failure 1af86ebfa128ce62722c1385bcf91729349b907c

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/289)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/295)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/275)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/218)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/218)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/284)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/289)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/287)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/293)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/287)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/293)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/312)
* try-community-x86_64-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/284)

**Comment by livecodepanos** [comment]
@livecode-vulcan retry 1af86eb

**Comment by livecode-vulcan** [comment]
😎 test success 1af86ebfa128ce62722c1385bcf91729349b907c

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/291)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/297)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/277)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/220)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/220)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/286)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/291)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/289)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/295)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/289)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/295)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/314)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/286)

**Pull request merged**